bee-api/bee-rest-api/src/endpoints/routes/api/v1/submit_message.rs

```diff
@@ -0,0 +1,231 @@
+use log::error;
+use serde_json::Value as JsonValue;
+use tokio::sync::mpsc;
+use warp::{http::StatusCode, reject, Filter, Rejection, Reply};
+
+use std::{convert::TryFrom, net::IpAddr};
+
+fn path() -> impl Filter<Extract = (), Error = Rejection> + Clone {
+    super::path().and(warp::path("messages")).and(warp::path::end())
+}
+
+pub(crate) fn filter<B: StorageBackend>(
+    public_routes: Vec<String>,
+    allowed_ips: Vec<IpAddr>,
+    tangle: ResourceHandle<MsTangle<B>>,
+    message_submitter: mpsc::UnboundedSender<MessageSubmitterWorkerEvent>,
+    network_id: NetworkId,
+    rest_api_config: RestApiConfig,
+    protocol_config: ProtocolConfig,
+) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
+    self::path()
+        .and(warp::post())
+        .and(has_permission(ROUTE_SUBMIT_MESSAGE, public_routes, allowed_ips))
+        .and(warp::body::json())
+        .and(with_tangle(tangle))
+        .and(with_message_submitter(message_submitter))
+        .and(with_network_id(network_id))
+        .and(with_rest_api_config(rest_api_config))
+        .and(with_protocol_config(protocol_config))
+        .and_then(submit_message)
+}
+
+pub(crate) async fn submit_message<B: StorageBackend>(
+    value: JsonValue,
+    tangle: ResourceHandle<MsTangle<B>>,
+    message_submitter: mpsc::UnboundedSender<MessageSubmitterWorkerEvent>,
+    network_id: NetworkId,
+    rest_api_config: RestApiConfig,
+    protocol_config: ProtocolConfig,
+) -> Result<impl Reply, Rejection> {
+    let network_id_v = &value["networkId"];
+    let parents_v = &value["parentMessageIds"];
+    let payload_v = &value["payload"];
+    let nonce_v = &value["nonce"];
+
+    // parse the values, take care of missing fields, build the message and submit it to the node for further
+    // processing
+
+    let network_id = if network_id_v.is_null() {
+        network_id.1
+    } else {
+        network_id_v
+            .as_str()
+            .ok_or_else(|| {
+                reject::custom(CustomRejection::BadRequest(
+                    "invalid network id: expected an u64-string".to_string(),
+                ))
+            })?
+            .parse::<u64>()
+            .map_err(|_| {
+                reject::custom(CustomRejection::BadRequest(
+                    "invalid network id: expected an u64-string".to_string(),
+                ))
+            })?
+    };
+
+    let parents: Vec<MessageId> = if parents_v.is_null() {
+        tangle.get_messages_to_approve().await.ok_or_else(|| {
+            reject::custom(CustomRejection::ServiceUnavailable(
+                "can not auto-fill parents: no tips available".to_string(),
+            ))
+        })?
+    } else {
+        let array = parents_v.as_array().ok_or_else(|| {
+            reject::custom(CustomRejection::BadRequest(
+                "invalid parents: expected an array of message ids".to_string(),
+            ))
+        })?;
+        let mut message_ids = Vec::new();
```

Comment:
I'd use `Vec::with_capacity` to avoid multiple allocations

Reply:
fixed in cpt-2